File: vripper-server/src/main/java/tn/mnlr/vripper/web/wsendpoints/DataBroadcast.java
package tn.mnlr.vripper.web.wsendpoints;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;
import reactor.core.Disposable;
import tn.mnlr.vripper.event.Event;
import tn.mnlr.vripper.event.EventBus;
import tn.mnlr.vripper.jpa.domain.Image;
import tn.mnlr.vripper.services.DataService;
import tn.mnlr.vripper.services.domain.DownloadSpeed;
import tn.mnlr.vripper.services.domain.GlobalState;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import java.time.Duration;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@Slf4j
public class DataBroadcast {

  private final SimpMessagingTemplate template;
  private final DataService dataService;
  private final EventBus eventBus;

  private Disposable disposable;

  @Autowired
  public DataBroadcast(SimpMessagingTemplate template, DataService dataService, EventBus eventBus) {
    this.template = template;
    this.dataService = dataService;
    this.eventBus = eventBus;
  }

  @PostConstruct
  private void run() {

    disposable =
        eventBus
            .flux()
            .buffer(Duration.of(500, ChronoUnit.MILLIS))
            .subscribe(
                data -> {
                  Map<Event.Kind, List<Event<?>>> eventMap =
                      data.stream().collect(Collectors.groupingBy(Event::getKind));

                  eventMap.forEach(
                      (kind, eventList) -> {
                        switch (kind) {
                          case POST_UPDATE:
                          case METADATA_UPDATE:
                            template.convertAndSend(
                                "/topic/posts",
                                eventList.stream()
                                    .map(e -> ((Long) e.getData()))
                                    .distinct()
                                    .map(dataService::findById)
                                    .filter(Optional::isPresent)
                                    .map(Optional::get)
                                    .collect(Collectors.toUnmodifiableSet()));

                            break;
                          case POST_REMOVE:
                            template.convertAndSend(
                                "/topic/posts/deleted",
                                eventList.stream()
                                    .map(e -> ((String) e.getData()))
                                    .collect(Collectors.toUnmodifiableSet()));
                            break;
                          case IMAGE_UPDATE:
                            eventList.stream()
                                .map(e -> ((Long) e.getData()))
                                .distinct()
                                .map(dataService::findImageById)
                                .filter(Optional::isPresent)
                                .map(Optional::get)
                                .collect(Collectors.groupingBy(Image::getPostId))
                                .forEach(
                                    (postId, images) ->
                                        template.convertAndSend("/topic/images/" + postId, images));
                            break;

                          case QUEUED_UPDATE:
                            template.convertAndSend(
                                "/topic/queued",
                                eventList.stream()
                                    .map(e -> ((Long) e.getData()))
                                    .distinct()
                                    .map(dataService::findQueuedById)
                                    .filter(Optional::isPresent)
                                    .map(Optional::get)
                                    .collect(Collectors.toUnmodifiableSet()));
                            break;
                          case QUEUED_REMOVE:
                            template.convertAndSend(
                                "/topic/queued/deleted",
                                eventList.stream()
                                    .map(e -> ((String) e.getData()))
                                    .collect(Collectors.toUnmodifiableSet()));
                            break;
                          case LOG_EVENT_UPDATE:
                            template.convertAndSend(
                                "/topic/events",
                                eventList.stream()
                                    .map(e -> ((Long) e.getData()))
                                    .distinct()
                                    .map(dataService::findEventById)
                                    .filter(Optional::isPresent)
                                    .map(Optional::get)
                                    .collect(Collectors.toUnmodifiableSet()));
                            break;
                          case LOG_EVENT_REMOVE:
                            template.convertAndSend(
                                "/topic/events/deleted",
                                eventList.stream()
                                    .map(e -> ((Long) e.getData()))
                                    .collect(Collectors.toUnmodifiableSet()));
                            break;
                          case VG_USER:
                            eventList.stream()
                                .map(e -> new DataController.LoggedUser((String) e.getData()))
                                .distinct()
                                .forEach(user -> template.convertAndSend("/topic/user", user));
                            break;
                          case GLOBAL_STATE:
                            eventList.stream()
                                .map(e -> ((GlobalState) e.getData()))
                                .distinct()
                                .forEach(
                                    globalState ->
                                        template.convertAndSend(
                                            "/topic/download-state", globalState));
                            break;
                          case BYTES_PER_SECOND:
                            eventList.stream()
                                .map(e -> new DownloadSpeed((Long) e.getData()))
                                .distinct()
                                .forEach(speed -> template.convertAndSend("/topic/speed", speed));

                            break;
                        }
                      });
                });
  }

  @PreDestroy
  private void destroy() {
    if (disposable != null) {
      disposable.dispose();
    }
  }
}

File: vripper-web/src/main/kotlin/me/mnlr/vripper/web/wsendpoints/DataBroadcast.kt
package me.mnlr.vripper.web.wsendpoints

import jakarta.annotation.PostConstruct
import jakarta.annotation.PreDestroy
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Service
import reactor.core.Disposable
import me.mnlr.vripper.entities.ImageDownloadState
import me.mnlr.vripper.event.Event
import me.mnlr.vripper.event.EventBus
import me.mnlr.vripper.formatSI
import me.mnlr.vripper.model.GlobalState
import me.mnlr.vripper.repositories.ImageRepository
import me.mnlr.vripper.repositories.LogEventRepository
import me.mnlr.vripper.repositories.PostDownloadStateRepository
import me.mnlr.vripper.repositories.ThreadRepository
import java.time.Duration
import java.time.temporal.ChronoUnit

@Service
class DataBroadcast constructor(
    private val template: SimpMessagingTemplate,
    private val eventBus: EventBus,
    private val postDownloadStateRepository: PostDownloadStateRepository,
    private val imageRepository: ImageRepository,
    private val threadRepository: ThreadRepository,
    private val logEventRepository: LogEventRepository
) {
    var globalState: GlobalState = GlobalState(0, 0, 0, "", "")
    private var disposable: Disposable? = null

    @PostConstruct
    private fun run() {
        disposable = eventBus
            .flux()
            .buffer(Duration.of(500, ChronoUnit.MILLIS))
            .subscribe { events: List<Event<*>> ->
                events.groupBy { it.kind }.forEach { (kind: Event.Kind, eventList: List<Event<*>>) ->
                    when (kind) {
                        Event.Kind.POST_UPDATE, Event.Kind.METADATA_UPDATE -> template.convertAndSend(
                            "/topic/posts",
                            eventList
                                .asSequence()
                                .map { (_, data): Event<*> -> data as Long }
                                .distinct()
                                .map { postDownloadStateRepository.findById(it) }
                                .filter { it.isPresent }
                                .map { it.get() }
                                .toList())

                        Event.Kind.POST_REMOVE -> template.convertAndSend(
                            "/topic/posts/deleted",
                            eventList.map { (_, data): Event<*> -> data as String })

                        Event.Kind.IMAGE_UPDATE -> eventList
                            .asSequence()
                            .map { (_, data): Event<*> -> data as Long }
                            .distinct()
                            .map { imageRepository.findById(it) }
                            .filter { it.isPresent }
                            .map { it.get() }
                            .groupBy { it.postId }
                            .forEach { (postId: String, images: List<ImageDownloadState>) ->
                                template.convertAndSend(
                                    "/topic/images/$postId", images
                                )
                            }

                        Event.Kind.THREAD_UPDATE -> template.convertAndSend(
                            "/topic/threads",
                            eventList
                                .asSequence()
                                .map { (_, data): Event<*> -> data as Long }
                                .distinct()
                                .map { threadRepository.findById(it) }
                                .filter { it.isPresent }
                                .map { it.get() }
                                .toList())

                        Event.Kind.THREAD_REMOVE -> template.convertAndSend(
                            "/topic/threads/deleted",
                            eventList
                                .map { (_, data): Event<*> -> data as String })

                        Event.Kind.THREAD_CLEAR -> template.convertAndSend(
                            "/topic/threads/deletedAll",
                            eventList
                                .map { true })

                        Event.Kind.LOG_EVENT_UPDATE -> template.convertAndSend(
                            "/topic/logs",
                            eventList
                                .asSequence()
                                .map { (_, data): Event<*> -> data as Long }
                                .distinct()
                                .map { logEventRepository.findById(it) }
                                .filter { it.isPresent }
                                .map { it.get() }
                                .toList())

                        Event.Kind.LOG_EVENT_REMOVE -> template.convertAndSend(
                            "/topic/logs/deleted",
                            eventList
                                .map { (_, data): Event<*> -> data as Long })

                        Event.Kind.VG_USER -> eventList
                            .distinct()
                            .forEach {
                                globalState = globalState.copy(loggedUser = it.data as String)
                                template.convertAndSend("/topic/state", globalState)
                            }
                        Event.Kind.DOWNLOAD_STATUS -> eventList
                            .distinct()
                            .forEach {
                                val state = it.data as GlobalState
                                globalState = globalState.copy(running = state.running, remaining = state.remaining, error = state.error)
                                template.convertAndSend("/topic/state", globalState)
                            }
                        Event.Kind.BYTES_PER_SECOND -> eventList
                            .distinct()
                            .forEach {
                                globalState = globalState.copy(downloadSpeed = (it.data as Long).formatSI())
                                template.convertAndSend("/topic/state", globalState)
                            }

                        else -> {}
                    }
                }
            }
    }

    @PreDestroy
    private fun destroy() {
        if (disposable != null) {
            disposable!!.dispose()
        }
    }
}